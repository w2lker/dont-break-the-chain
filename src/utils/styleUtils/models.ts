export type cssAttribute = 'unset' | 'null' | number | string | undefined | null;

type TArrayWithLength<TItem, TLength extends number> = [TItem, ...TItem[]] & { length: TLength };
export type BoundAttributeArray = TArrayWithLength<string, 1| 2 | 3 | 4>;
