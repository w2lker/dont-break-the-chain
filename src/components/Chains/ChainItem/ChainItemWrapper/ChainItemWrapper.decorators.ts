import withStyles from '@material-ui/core/styles/withStyles';

import ChainItemWrapper from './ChainItemWrapper';
import chainItemWrapperStyles from './ChainItemWrapper.styles';

// @ts-ignore
const styledComponent = withStyles(chainItemWrapperStyles)(ChainItemWrapper);

export default styledComponent;
