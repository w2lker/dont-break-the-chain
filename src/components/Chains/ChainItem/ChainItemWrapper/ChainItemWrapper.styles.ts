import { styleUtils } from '../../../../utils/styleUtils';

const wrapperAsideBorders = '40px solid transparent';
const wrapperTopBorders = '10px solid #ffffff';

const chainItemWrapperStyles = {
  wrapper: {
    width: 80,
    margin: '0 45px',
    textAlign: 'center',
    position: 'relative',
  },
  wrapperStart: {
    '&:before': {
      content: '""',
      display: 'block',
      ...styleUtils.position('absolute', {
        top: 0,
        right: 0,
        left: 0,
      }),
      borderLeft: wrapperAsideBorders,
      borderRight: wrapperAsideBorders,
      borderTop: wrapperTopBorders,
    },
  },
  wrapperEnd: {
    '&:before': {
      content: '""',
      display: 'block',
      ...styleUtils.position('absolute', {
        right: 0,
        bottom: 0,
      }),
      borderBottom: wrapperTopBorders,
      borderLeft: wrapperAsideBorders,
    },
    '&.after': {
      content: '""',
      display: 'block',
      ...styleUtils.position('absolute', {
        bottom: 0,
        left: 0,
      }),
      borderBottom: wrapperTopBorders,
      borderRight: wrapperAsideBorders,
    },
  },
  wrapperIncompleteToday: {
    zIndex: 0,
    '&:before': {
      content: '""',
      display: 'block',
      ...styleUtils.position('absolute', {
        right: 0,
        bottom: 0,
      }),
      borderBottom: wrapperTopBorders,
      borderLeft: wrapperAsideBorders,
    },
    '&:after': {
      content: '""',
      display: 'block',
      ...styleUtils.position('absolute', {
        bottom: 0,
        left: 0,
      }),
      borderBottom: wrapperTopBorders,
      borderRight: wrapperAsideBorders,
    },
  },
  wrapperComplete: {
    position: 'relative',
  },
  customBackground: {
    ...styleUtils.position('absolute', {
      top: -12,
      right: 0,
      bottom: 0,
      left: 0,
    }),
    opacity: 0.3,
  },
  triangleItem: {
    zIndex: 10,
    ...styleUtils.position('absolute', {
      top: -12,
    }),
    borderBottom: '12px solid',
    borderBottomColor: 'inherit',
  },
  triangleLeft: {
    left: 0,
    borderRight: '42px solid transparent',
  },
  triangleRight: {
    right: 0,
    borderLeft: '42px solid transparent',
  },
};

export default chainItemWrapperStyles;
