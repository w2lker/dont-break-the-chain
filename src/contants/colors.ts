import colorBlue from '@material-ui/core/colors/blue';
import colorBrown from '@material-ui/core/colors/brown';
import colorCyan from '@material-ui/core/colors/cyan';
import colorDeepPurple from '@material-ui/core/colors/deepPurple';
import colorGreen from '@material-ui/core/colors/green';
import colorIndigo from '@material-ui/core/colors/indigo';
import colorOrange from '@material-ui/core/colors/orange';
import colorPink from '@material-ui/core/colors/pink';
import colorPurple from '@material-ui/core/colors/purple';
import colorTeal from '@material-ui/core/colors/teal';

export type colorsKeys = 'pink' | 'purple' | 'deepPurple' | 'indigo' | 'blue' | 'cyan' | 'teal' | 'green' | 'orange' | 'brown';

export const colors: {[color: string]: any}  = {
  blue: colorBlue,
  brown: colorBrown,
  pink: colorPink,
  cyan: colorCyan,
  deepPurple: colorDeepPurple,
  indigo: colorIndigo,
  green: colorGreen,
  orange: colorOrange,
  purple: colorPurple,
  teal: colorTeal,
};
