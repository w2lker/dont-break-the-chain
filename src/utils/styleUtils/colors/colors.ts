import {bonWarning} from '../helpers/warning/warning';

export interface IRgbColor {
  red: number;
  green: number;
  blue: number;
}

const black: IRgbColor = {
  red: 0,
  green: 0,
  blue: 0,
};

function regularHexToRgb(hex: string): IRgbColor {
  const [red, green, blue] = [0, 2, 4]
    .map(offset => hex.substring(offset, offset + 2))
    .map(hexCh => parseInt(hexCh, 16));
  return { red, green, blue };
}

export function hexToRgbMapped(hex?: string): IRgbColor {
  const warnType = 'HEX to RGB converter';
  if (!hex) {
    return black;
  }
  if (hex.length === 1) {
    return regularHexToRgb(hex.repeat(6));
  }

  if (hex.length === 2) {
    return regularHexToRgb(hex.repeat(3));
  }

  if (hex.length === 3) {
    const [r, g, b] = hex.split('') as string[];
    return regularHexToRgb(`${r}${r}${g}${g}${b}${b}`);
  }

  if (hex.length === 4 || hex.length === 5) {
    const hex3 = hex.slice(0, 3);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex3} value`, warnType);
    return hexToRgbMapped(hex3);
  }

  if (hex.length > 6) {
    const hex6 = hex.slice(0, 6);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex6} value`, warnType);
    return regularHexToRgb(hex6);
  }

  return regularHexToRgb(hex);
}
