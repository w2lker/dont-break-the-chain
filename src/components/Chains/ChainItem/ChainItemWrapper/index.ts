import ChainItemWrapper from "./ChainItemWrapper";
export default ChainItemWrapper;