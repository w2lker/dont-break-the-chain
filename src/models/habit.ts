import {List} from 'immutable';
import {colorsKeys} from "../contants/colors";

export const habitStatus = {
  incomplete: 0,
  complete: 1,
  paused: 2,
  start: 3,
  end: 4,
  incompleteToday: 5
};

export interface IHabitDate {
  date?: any;
  status?: string;
  score?: number;
}

export interface IHabit {
  id?: number;
  name?: string;
  color?: colorsKeys;
  icon?: string;
  currentChain?: number;
  longestChain?: number;
  dates?: List<IHabitDate>;
}