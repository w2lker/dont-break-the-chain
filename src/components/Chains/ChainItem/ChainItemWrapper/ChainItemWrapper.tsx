import React from "react";
import classNames from 'classnames';

import {colors, colorsKeys} from "../../../../contants/colors";
import {habitStatus} from "../../../../models/habit";

export interface IChainItemWrapper {
  status: number,
  color: colorsKeys,
}

const ChainItemWrapper: React.FC<IChainItemWrapper> = (props) => {
  const {status} = props;
  const isHabitIncompleteToday = status === habitStatus.incompleteToday;

  const wrapperClass = classNames({
    'habit-chain-wrapper': true,
    'incomplete': status === habitStatus.incomplete,
    'complete' : status === habitStatus.complete,
    'paused': status === habitStatus.paused,
    'start': status === habitStatus.start,
    'end': status === habitStatus.end,
    'incomplete-today': isHabitIncompleteToday
  });

  const color = colors[props.color || 'blue'][100];

  const wrapperStyle = {
    backgroundColor: !isHabitIncompleteToday ? color : null
  };

  return (
    <div
      className={wrapperClass}
      style={wrapperStyle}
    >
      {isHabitIncompleteToday && (
        <React.Fragment>
          <div
            className='custom-background'
            style={{backgroundColor: color}}
          />
          <span className='triangleLeft'/>
          <span className='triangleRight'/>
        </React.Fragment>
      )}
      {props.children}
    </div>
  )
};

ChainItemWrapper.defaultProps = {
  status: habitStatus.incomplete,
};

export default ChainItemWrapper;