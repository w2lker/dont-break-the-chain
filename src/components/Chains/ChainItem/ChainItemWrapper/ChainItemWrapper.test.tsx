import React from "react";
import { shallow, mount, render } from "enzyme";
import ChainItemWrapper, {IChainItemWrapper} from "./ChainItemWrapper";
import {colors} from "../../../../contants/colors";
import {habitStatus} from "../../../../models/habit";
import {arrayWithout} from "../../../../utils/testUtils";

describe('Chain Item Wrapper', () => {
  const emptyProps: IChainItemWrapper = {
    // @ts-ignore
    status: null,
    // @ts-ignore
    color: null
  };

  const wrapperClassNames = ['incomplete', 'complete', 'paused', 'start', 'end', 'incomplete-today'];
  const triangleClasses = ['.triangleLeft', '.triangleRight', '.custom-background'];

  it('match snapshot', () => {
    const snapshotProps: IChainItemWrapper = {
      status: habitStatus.incompleteToday,
      color: "deepPurple",
    };
    const component = mount(<ChainItemWrapper {...snapshotProps} />);
    expect(component.debug()).toMatchSnapshot();
  });
  describe('Renders with empty props', () => {
    const component = shallow(<ChainItemWrapper {...emptyProps} />);
    it('renders wrapper', () => {
      expect(component.find('div.habit-chain-wrapper').exists()).toBeTruthy();
    });
    it("doesn't render wrapper classes", () => {
      const habitWrapper = component.find('.habit-chain-wrapper');
      wrapperClassNames.forEach( className => {
        expect(habitWrapper.hasClass(className)).toBeFalsy();
      });
    });
    it('skips triangles render', () => {
      triangleClasses.forEach( className => {
        expect(component.find(className).exists()).toBeFalsy();
      });
    });
    it('applies default color value', () => {
        const defaultStyle = {
          backgroundColor: colors['blue'][100],
        };
        expect(component.find('.habit-chain-wrapper').props().style).toEqual(defaultStyle);
    });
  });
  describe('Renders incomplete today status', () => {
    const newProps: IChainItemWrapper = {
      status: habitStatus.incompleteToday,
      color: "green",
    };
    const component = shallow(<ChainItemWrapper {...newProps} />);
    it('renders wrapper', () => {
      expect(component.find('div.habit-chain-wrapper').exists()).toBeTruthy();
      expect(component.find('.habit-chain-wrapper').hasClass('incomplete-today')).toBeTruthy();
    });
    it("doesn't render wrapper classes", () => {
      const habitWrapper = component.find('.habit-chain-wrapper');
      const incompleteClasses = arrayWithout(wrapperClassNames, 'incomplete-today');
      incompleteClasses.forEach( className => {
        expect(habitWrapper.hasClass(className)).toBeFalsy();
      });
    });
    it("executes triangles render", () => {
      triangleClasses.forEach( className => {
        expect(component.find(className).exists()).toBeTruthy();
      });
    });
  });
  it('Renders other statuses', () => {
      const getProps = (status: number): IChainItemWrapper => ({
        status,
        color: 'green'
      });

      const statuses = ['incomplete', 'complete', 'paused', 'start', 'end'];
      statuses.forEach( status => {
        // @ts-ignore
        const component = shallow(<ChainItemWrapper {...getProps(habitStatus[status])} />);
        const otherClasses = arrayWithout(wrapperClassNames, status);
        otherClasses.forEach( className => {
          expect(component.find('.habit-chain-wrapper').hasClass(className)).toBeFalsy();
        });
        triangleClasses.forEach( className => {
          expect(component.find(className).exists()).toBeFalsy();
        });
      });
    })
});