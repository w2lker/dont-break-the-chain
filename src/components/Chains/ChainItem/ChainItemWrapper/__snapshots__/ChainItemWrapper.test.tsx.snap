// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chain Item Wrapper match snapshot 1`] = `
"<ChainItemWrapper status={5} color=\\"deepPurple\\">
  <div className=\\"habit-chain-wrapper incomplete-today\\" style={{...}}>
    <div className=\\"custom-background\\" style={{...}} />
    <span className=\\"triangleLeft\\" />
    <span className=\\"triangleRight\\" />
  </div>
</ChainItemWrapper>"
`;
